feat(fileio): Introduce file io performance problem

wmak opened the pull request:
- Introduce the file io on main thread performance problem. Not hooking up the options or anything yet. Just want to get the main core of the code ready to go first
- I think this won't actually create any of the file io perf problems yet since the option is only added to default.py with 0
- Depends on https://github.com/getsentry/rfcs/pull/36

wmak committed:
feat(fileio): Introduce file io performance problem

- Introduce the file io on main thread performance problem. Not hooking
  up the options or anything yet. Just want to get the main core of the
  code ready to go first
- I think this won't actually create any of the file io perf problems
  yet since the option is only added to default.py with 0

wmak committed:
lets be safe, remove the option entirely

wmak committed:
Removing projectsetting since we don't really need it

wmak committed:
fix: handle data being potentially None

romtsn commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -1205,6 +1214,52 @@ class NPlusOneDBSpanDetectorExtended(NPlusOneDBSpanDetector):
+            op, span_id, op_prefix, span_duration, settings = settings_for_span
+            if span_duration.total_seconds() * 1000 > settings["duration_threshold"]:
+                call_stack = ".".join(
+                    [item["function"] for item in span.get("data", {}).get("call_stack", [])]
```
I guess we should also consider using `filename` or `module` for the fingerprint, because different classes can have identical function names in theory

mjq-sentry commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -1205,6 +1214,52 @@ class NPlusOneDBSpanDetectorExtended(NPlusOneDBSpanDetector):
+                    desc=span.get("description", ""),
+                    parent_span_ids=[],
+                    type=GroupType.PERFORMANCE_FILE_IO_MAIN_THREAD,
+                    cause_span_ids=[span.get("span_id", None)],
```
I think `cause_span_ids  should be empty in this case. "Cause" is for related spans that caused the bad spans, whereas `offender_span_ids` are the actual bad spans. (Cause might not make sense for anything other than N+1 tbh).

mjq-sentry commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -1205,6 +1214,52 @@ class NPlusOneDBSpanDetectorExtended(NPlusOneDBSpanDetector):
+        if data is None:
+            return False
+        # doing is True since the value can be any type
+        return data.get("blocked_ui_thread", False) is True
```
Should there be any restrictions on the span op here as well, or will all spans with `blocked_ui_thread` always be file I/O? Based on the name I could see it getting used on other operations down the line too.

wmak replied:
Hmm I'll add a prefix check for `file` good callout :+1:

romtsn replied:
In the rfc https://github.com/getsentry/rfcs/pull/36 we say that this flag works only for auto-instrumented spans, and when we know for sure that the span spent its entire time on the main/ui thread. We also mentioned that we can potentially expand it to db and http ops.

So I'm not sure if we should make the detector generic already now, so we can add new perf issues without any changes here, or maintain the list of allowed ops whenever we decide to add something new

wmak replied:
Changing the code here in the detector is pretty easy. But I'll remove the op check for now. We aren't actually creating the issues yet so I'd rather collect too many metrics than too few

wmak committed:
ref: Address PR comments

- Remove cause_span_ids
- Check for op prefix is file
- Include module in the call stack hashing

wmak committed:
ref: remove the op check

gggritso commented on tests/sentry/utils/performance_issues/test_performance_detection.py:
```diff
@@ -980,6 +983,26 @@ def test_detects_consecutive_db_spans(self):
             ]
         )
 
+    def test_detects_file_io_main_thread(self):
+        file_io_event = EVENTS["file-io-on-main-thread"]
+        sdk_span_mock = Mock()
+
+        _detect_performance_problems(file_io_event, sdk_span_mock)
+
+        assert sdk_span_mock.containing_transaction.set_tag.call_count == 5
+        sdk_span_mock.containing_transaction.set_tag.assert_has_calls(
+            [
+                call("_pi_all_issue_count", 1),
+                call("_pi_sdk_name", "sentry.java.android.timber"),
+                call("_pi_transaction", "c119e45a9d724b1891df4651ebf9e6db"),
+                call(
+                    "_pi_file_io_main_thread_fp",
+                    "1-GroupType.PERFORMANCE_FILE_IO_MAIN_THREAD-153198dd61706844cf3d9a922f6f82543df8125f",
+                ),
+                call("_pi_file_io_main_thread", "054ba3a374d543eb"),
+            ]
+        )
+
```
This style of test is a vestige of the before times, when we didn't return `PerformanceProblem` instances from the detector! Would be even nicer here to do something like this:

```python
problems = _detect_performance_problems(EVENTS["file-io-on-main-thread"], sdk_span_mock)

assert problems == [
  PerformanceProblem(
    fingerprint="yadda yadda",
    op="etc"
]
```

like `test_n_plus_one_extended_detection_no_parent_span`

wmak replied:
Nice, I thought it was strange that we just check the tag mock but wanted to match existing tests

wmak replied:
Nevermind, can't do this test format until I add the option which based on Matt's comment earlier i've removed for now

gggritso replied:
Ack, I thought that might be the case, sorry I didn't check ahead of time!

gggritso commented on src/sentry/utils/performance_issues/performance_detection.py:
```diff
@@ -1205,6 +1216,55 @@ class NPlusOneDBSpanDetectorExtended(NPlusOneDBSpanDetector):
+                        for item in span.get("data", {}).get("call_stack", [])
+                    ]
+                ).encode("utf8")
+                hashed_stack = hashlib.sha1(call_stack).hexdigest()
+                fingerprint = f"1-{GroupType.PERFORMANCE_FILE_IO_MAIN_THREAD}-{hashed_stack}"
```
Nit, but it'd be nice to put this in a `_fingerprint` method so it matches the other detectors! Maybe we should make an abstractmethod for this so it's more clear 🤔

gggritso commented on fixtures/events/performance_problems/file-io-on-main-thread.json:
```diff
@@ -0,0 +1 @@
+{"event_id":"c119e45a9d724b1891df4651ebf9e6db","project":5428559,"release":"io.sentry.samples.android@1.1.0+2","dist":"2","platform":"java","message":"","datetime":"2022-11-21T11:57:38.806589+00:00","tags":[["device","Android SDK built for x86"],["device.family","Android"],["environment","debug"],["isSideLoaded","true"],["level","info"],["os","Android 10"],["os.name","Android"],["os.rooted","no"] ...
```
Pretty-printing this is a good idea IMO. Eventually someone will want to figure out why a certain test case does or does not work, and it's good to be able to read through the file to see what's there. Also, it'll make apparent which data is not required for detector, or needs to be anonymized, or what have you

gggritso commented on tests/sentry/utils/performance_issues/test_performance_detection.py:
```diff
@@ -980,6 +983,26 @@ def test_detects_consecutive_db_spans(self):
             ]
         )
 
+    def test_detects_file_io_main_thread(self):
```
Would not be mad at a second test for the negative case! e.g., file IO _not_ on the main thread doesn't create a performance problem, or something similar

gggritso reviewed (commented):
Very awesome! Nits only 👏🏻

wmak committed:
address pr commetns

gggritso reviewed (approved):
B e a utiful
